File: Unit/Unit.cpp
#include <iostream>
#include "Unit.h"

using namespace std;

void Unit::ensureIsAlive() {
    if ( hitPoints == 0 ) {
        throw UnitIsDeadException();
    }
}

Unit::Unit(const std::string& name, int hp, int dmg) {
	this->name = name;
	this->hitPoints = hp;
	this->hitPointsLimit = hp;
	this->damage = dmg;
}

Unit::~Unit() {}

int Unit::getDamage() const {
	return damage;
}

int Unit::getHitPoints() const {
	return hitPoints;
}

int Unit::getHitPointsLimit() const {
	return hitPointsLimit;
}

const std::string& Unit::getName() const {
	return name;
}

void Unit::addHitPoints(int hp) {
	ensureIsAlive();

	if ( hitPointsLimit - hitPoints < hp ) {
		hitPoints = hitPointsLimit;

		return;
	}
	hitPoints += hp;
}

void Unit::takeDamage(int dmg) {
	ensureIsAlive();
	hitPoints -= dmg;

	if ( hitPoints <= 0 ) {
		hitPoints = 0;
	}
}

void Unit::attack(Unit& enemy) {
	ensureIsAlive();
	enemy.ensureIsAlive();
	enemy.takeDamage(damage);
	cout << getName() << " taked " << damage << " damage." << endl;

	enemy.counterAttack(*this);
}

void Unit::counterAttack(Unit& enemy) {
	ensureIsAlive();

	enemy.takeDamage(damage/2);

	cout << getName() << " counterattacking and taked " << damage/2 << " damage." << endl;
}

ostream& operator<<(ostream& out, const Unit& unit) {
	out << "Unit name: " << unit.getName() << endl;
	out << "Health: " << unit.getHitPointsLimit() << "/" << unit.getHitPoints() << endl;
	out << "Unit's damage: " << unit.getDamage() << endl;

	return out;
}

File: Java/Unit/Unit.java
public class Unit implements Cloneable{
    private String name;
    private int hitPoints;
    private int hitPointsLimit;
    private int damage;

    public Unit(String name, int hitPoint, int damage) {
        this.name = name;
        this.hitPoints = hitPoint;
        this.hitPointsLimit = hitPoint;
        this.damage = damage;
    }

    public void addHitPoints(int hp) throws UnitIsDeadException {
        ensureIsAlive();

        if (hitPointsLimit - hitPoints < hp) {
            hitPoints = hitPointsLimit;

            return;
        }
        hitPoints += hp;
    }

    public void takeDamage(int dmg) throws UnitIsDeadException {
        ensureIsAlive();

        hitPoints -= dmg;

        if ( hitPoints <= 0 ) {
            hitPoints = 0;
        }
    }

    public void attack(Unit enemy) throws UnitIsDeadException {
        ensureIsAlive();

        enemy.takeDamage(damage);
        System.out.println(getName() + " taked " + damage + " damage.\n");

        enemy.counterAttack(this);
    }

    public void counterAttack(Unit enemy) throws UnitIsDeadException {
        ensureIsAlive();

        enemy.takeDamage(damage/2);

        System.out.println(getName() + " counterattacking and taked " + damage/2 + " damage.\n");
    }

    public String toString() {
        StringBuffer out = new StringBuffer();

        out.append("Unit: " + getName() + "\n");
        out.append("Health: " + getHitPointsLimit() + "/" + getHitPoints() + "\n");
        out.append("Unit's damage: " + getDamage() + "\n");

        return out.toString();
    }

    public Unit clone() throws CloneNotSupportedException{
        return (Unit) super.clone();
    }

    public int getDamage() {
        return damage;
    }

    public int getHitPoints() {
        return hitPoints;
    }

    public int getHitPointsLimit() {
        return hitPointsLimit;
    }

    public String getName() {
        return name;
    }

    private void ensureIsAlive() throws UnitIsDeadException{
        if ( hitPoints == 0 ) {
            throw new UnitIsDeadException();
        }
    }
}
